core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2098,31 +2101,40 @@ class Log(@volatile var dir: File,
     logString.toString
   }
 
+  private def removeAndDeleteSegments(segments: Iterable[LogSegment], asyncDelete: Boolean = true): Unit = {
+    // As most callers hold an iterator into the `segments` collection and `removeAndDeleteSegment` mutates it by
+    // removing the deleted segment, we should force materialization of the iterator here, so that results of the
+    // iteration remain valid and deterministic.
+    segments.toList.foreach { segment =>
+      removeAndDeleteSegment(segment, asyncDelete)
+    }
+  }
+
   /**
-   * This method performs an asynchronous log segment delete by doing the following:
+   * This method deletes the given log segment by doing the following:
    * <ol>
    *   <li>It removes the segment from the segment map so that it will no longer be used for reads.
    *   <li>It renames the index and log files by appending .deleted to the respective file name
-   *   <li>It schedules an asynchronous delete operation to occur in the future
+   *   <li>It can either schedule an asynchronous delete operation to occur in the future or perform the deletion synchronously
    * </ol>
-   * This allows reads to happen concurrently without synchronization and without the possibility of physically
-   * deleting a file while it is being read from.
+   * Asynchronous deletion allows reads to happen concurrently without synchronization and without the possibility of
+   * physically deleting a file while it is being read.
    *
    * This method does not need to convert IOException to KafkaStorageException because it is either called before all logs are loaded
    * or the immediate caller will catch and handle IOException
    *
    * @param segment The log segment to schedule for deletion
+   * @param asyncDelete Whether the segment files should be deleted asynchronously
    */
-  private def deleteSegment(segment: LogSegment) {
-    info(s"Scheduling log segment [baseOffset ${segment.baseOffset}, size ${segment.size}] for deletion.")
+  private def removeAndDeleteSegment(segment: LogSegment, asyncDelete: Boolean = true) {
```

Comment:
where is asyncDelete == false used ?

Reply:
https://github.com/apache/kafka/pull/7106/files#diff-ffa8861e850121997a534ebdde2929c6R1965